### PR TITLE
Added fast_mimic for faster mimic using more space

### PR DESCRIPTION
**Pull request by parkds**
Added fast_mimic argument (default value False for compatibility) to mimic algorithm. Modified eval_node_probs in opt_probs.py to compute the MI in vectorized form across all row pairs. (dropped scikit learn implementation of MI as it can only compute on a pair by pair basis)

fast mimic can be used by setting fast_mimic=True when calling the mimic algorithm.

### CODE REVIEW

**Comment by parkds** [comment]
modified log order to reduce floating point error. Previous error was on the order of e-13, now the error is on the order of e-15.

File: mlrose/opt_probs.py [hunk 1]
```diff
@@ -275,13 +275,69 @@ def __init__(self, length, fitness_fn, maximize=True, max_val=2):
...
         """Update probability density estimates.
         """
-        # Create mutual info matrix
-        mutual_info = np.zeros([self.length, self.length])
-        for i in range(self.length - 1):
-            for j in range(i + 1, self.length):
-                mutual_info[i, j] = -1 * mutual_info_score(
-                    self.keep_sample[:, i],
-                    self.keep_sample[:, j])
+        if (self.mimic_speed == False):
```
Comment (gkhayes) [inline mlrose/opt_probs.py]: `self.mimic_speed` hasn't previously been defined for this Class, so this line is causing errors when I run the unit tests on this code.

File: mlrose/opt_probs.py [hunk 2]
```diff
@@ -275,13 +275,69 @@ def __init__(self, length, fitness_fn, maximize=True, max_val=2):
...
+            # get length of the sample which survived from mimic iteration
+            len_sample_kept = self.keep_sample.shape[0]
+            # get the length of the bit sequence / problem size
+            len_prob = self.keep_sample.shape[1]
+
+            # Expand the matrices to so each row corresponds to a row by row combination of the list of samples
+            b = np.repeat(self.keep_sample, self.length).reshape(len_sample_kept, len_prob * len_prob)
+            d = np.hstack(([self.keep_sample] * len_prob))
+
+            # Compute the mutual information matrix in bulk, by iterating through the list of possible feature values ((max_val-1)^2).
```
Comment (gkhayes) [inline mlrose/opt_probs.py]: Ideally, I would like the code for this package to be PEP8 compliant, which means that all lines of code need to be < 79 characters long. I think there are a few lines here that exceed that limit.

Comment (parkds) [reply mlrose/opt_probs.py]: Updated length, variable names, as well as line spacing for PEP8 compliance

File: mlrose/opt_probs.py (hunk 1 above)
Comment (parkds) [reply mlrose/opt_probs.py]: Initialized self.mimic_speed = False in DiscreteOpt base class __init__ method

**Comment by parkds** [comment]
Hi @gkhayes! I made the requested changes to initialize mimic_speed and made other changes for pep compliance! Please let me know if any other changes are required!

**Comment by gkhayes** [comment]
Hi @parkds,

Thanks for making the changes. I have merged your pull request and your changes will flow through to the next PyPi release of mlrose.

Regards,

Genevieve.

**Comment by YMa508** [comment]
may I ask when is the next PyPi release?

**Comment by gkhayes** [comment]
Hi @YMa508,
mlrose v1.3.0 has just been released on PyPi today.
Regards,
Genevieve.

**Pull request merged**